src/box/user.cc

```diff
@@ -906,6 +906,28 @@ role_revoke(struct user *grantee, struct user *role)
 	return 0;
 }
 
+/**
+ * Check if a role is granted to a user or role with the given auth token.
+ */
+bool
+role_is_granted(struct user *role, uint8_t auth_token)
+{
+	/* Check if the role is granted directly. */
+	if (user_map_is_set(&role->users, auth_token))
+		return true;
+
+	/* Check if the role is granted transitively. */
+	struct user_map_iterator it;
+	user_map_iterator_init(&it, &role->users);
+	struct user *transitive_role;
+	while ((transitive_role = user_map_iterator_next(&it))) {
+		if (transitive_role->def->type == SC_ROLE &&
+		    role_is_granted(transitive_role, auth_token))
+			return true;
+	}
+	return false;
```

Comment:
I would use BFS instead of DFS here:
1. I believe it's a good practice to beware of recursion.
2. It would be consistent with implementation of `role_check` function.

Function `role_check` is a good example of BFS implementation for privileges.